contracts/solidity/NFTXYieldStakingZap.sol

```diff
@@ -0,0 +1,336 @@
+
+  function pairFor(address tokenA, address tokenB) internal view returns (address pair) {
+    (address token0, address token1) = sortTokens(tokenA, tokenB);
+    pair = address(uint160(uint256(keccak256(abi.encodePacked(
+      hex'ff',
+      sushiRouter.factory(),
+      keccak256(abi.encodePacked(token0, token1)),
+      hex'e18a34eb0e04b04f7a0ac29a6e80748dca96319b42c54d679cb821dca90c6303' // init code hash
+    )))));
+  }
+
+
+  /**
+   * @notice Returns sorted token addresses, used to handle return values from pairs sorted in
+   * this order.
+   */
+
+  function sortTokens(address tokenA, address tokenB) internal pure returns (address token0, address token1) {
+      require(tokenA != tokenB, 'UniswapV2Library: IDENTICAL_ADDRESSES');
+      (token0, token1) = tokenA < tokenB ? (tokenA, tokenB) : (tokenB, tokenA);
+      require(token0 != address(0), 'UniswapV2Library: ZERO_ADDRESS');
+  }
+
+
+  /**
+   * @notice Allows our owner to withdraw and tokens in the contract.
+   * 
+   * @param token The address of the token to be rescued
+   */
+
+  function rescue(address token) external onlyOwner {
+    if (token == address(0)) {
+      (bool success, ) = payable(msg.sender).call{value: address(this).balance}("");
+      require(success, "Address: unable to send value");
+    } else {
+      IERC20Upgradeable(token).safeTransfer(msg.sender, IERC20Upgradeable(token).balanceOf(address(this)));
+    }
+  }
+
+
+  /**
+   * @notice Swaps ERC20->ERC20 tokens held by this contract using a 0x-API quote.
+   *
+   * @param buyToken The `buyTokenAddress` field from the API response
+   * @param swapCallData The `data` field from the API response
+   */
+
+  function _fillQuote(
+    address buyToken,
+    bytes calldata swapCallData
+  ) internal returns (uint256) {
+      // Track our balance of the buyToken to determine how much we've bought.
+      uint256 boughtAmount = IERC20Upgradeable(buyToken).balanceOf(address(this));
+
+      // Give `swapTarget` an infinite allowance to spend this contract's `sellToken`.
+      // Note that for some tokens (e.g., USDT, KNC), you must first reset any existing
+      // allowance to 0 before being able to update it.
+      require(IERC20Upgradeable(address(WETH)).approve(swapTarget, type(uint256).max), 'Unable to approve contract');
+
+      // Call the encoded swap function call on the contract at `swapTarget`
+      (bool success,) = swapTarget.call(swapCallData);
+      require(success, 'SWAP_CALL_FAILED');
+
+      // Use our current buyToken balance to determine how much we've bought.
+      return IERC20Upgradeable(buyToken).balanceOf(address(this)) - boughtAmount;
+  }
+
+
+  /**
+   * @notice Maps a cached NFTX vault address against a vault ID.
+   * 
+   * @param vaultId The ID of the NFTX vault
+   */
+
+  function _vaultAddress(uint256 vaultId) internal returns (address) {
+    if (nftxVaultAddresses[vaultId] == address(0)) {
+      nftxVaultAddresses[vaultId] = nftxFactory.vault(vaultId);
```

Comment:
Not sure why this is needed when you can just use the factory no?

Reply:
Mentioned briefly in previous comment, but it is for gas savings for repeat calls

Reply:
Do you have a rough estimate on savings?